README.md
# php-color-output [![Build Status](https://travis-ci.org/jfcherng/php-color-output.svg?branch=master)](https://travis-ci.org/jfcherng/php-color-output)

![demo.gif](https://i.imgur.com/xMhYHjV.gif)

The above screenshot is the output of `demo.php`. See the [Example](#example) section.


# Installation

```
composer require jfcherng/php-color-output
```


# Available Colors

| Background   | Foreground   | Compound       | Special   | Alias         |
| ---          | ---          | ---            | ---       | ---           |
| b_black      | f_black      | f_dark_gray    | blink     | b (bold)      |
| b_blue       | f_blue       | f_light_blue   | bold      | blk (blink)   |
| b_cyan       | f_brown      | f_light_cyan   | dim       | h (hidden)    |
| b_green      | f_cyan       | f_light_green  | hidden    | rev (reverse) |
| b_light_gray | f_green      | f_light_purple | reset     | rst (reset)   |
| b_magenta    | f_light_gray | f_light_red    | reverse   | u (underline) |
| b_red        | f_normal     | f_white        | underline | -             |
| b_yellow     | f_purple     | f_yellow       | -         | -             |
| -            | f_red        | -              | -         | -             |


# Functions

```php
<?php

// a global alias to \Jfcherng\Color\Colorful::color
function str_color(string $str, $colors = [], bool $autoReset = true): string

// a global alias to \Jfcherng\Color\Colorful::noColor
function str_nocolor(string $str): string
```


# Methods

```php
<?php

/**
 * Make a string colorful.
 *
 * @param string       $str       the string
 * @param array|string $colors    the colors
 * @param bool         $autoReset automatically reset at the end of the string?
 *
 * @return string the colored string
 */
\Jfcherng\Color\Colorful::color(string $str, $colors = [], bool $autoReset = true): string

/**
 * Remove all colors from a string.
 *
 * @param string $str the string
 *
 * @return string the string without colors
 */
\Jfcherng\Color\Colorful::noColor(string $str): string
```


# Example

```php
<?php

include __DIR__ . '/vendor/autoload.php';

// colors in a string using a comma as the delimiter
echo str_color('foo', 'f_light_cyan, b_yellow');  // "\033[1;36;43mfoo\033[0m"

echo PHP_EOL;

// colors in an array
echo str_color('foo', ['f_white', 'b_magenta']); // "\033[1;37;45mfoo\033[0m"

echo PHP_EOL;

// do not auto reset color at the end of string
echo str_color('foo', ['f_red', 'b_green', 'b', 'blk'], false); // "\033[31;42;1;5mfoo"

// manually add color reset
echo str_color('', 'reset'); // "\033[0m"

echo PHP_EOL;

// remove all color codes from a string
echo str_nocolor("\033[31;42;5mfoo\033[0mbar"); // "foobar"

echo PHP_EOL;
```


Supporters <a href="https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=ATXYY9Y78EQ3Y" target="_blank"><img src="https://www.paypalobjects.com/en_US/i/btn/btn_donate_LG.gif" /></a>
==========

Thank you guys for sending me some cups of coffee.


CHANGELOG

VERSION 1  INIT
===============

   Version 1.0 - init
      27/11/2018 21:25  1.0.8  nits
         61d365e Update codes
         b6e69f0 Update README
         69d84bf Update deps
      04/09/2018 18:56  1.0.7  php-cs-fixer
         7dd1322 php-cs-fixer: enable native_constant_invocation and native_function_invocation
         1a1d921 Update deps
         1af0ae3 Rename CHANGELOG.md -> CHANGELOG
         47776c4 Update deps
         2de4f26 Update .travis.yml
      18/07/2018 08:47  1.0.6  Add composer.lock
         0281a77 Add composer.lock
      04/07/2018 15:33  1.0.5  nits
         fcecbfb Update .rmt.yml
      04/07/2018 15:06  1.0.4  initial release

tests/bootstrap.php
<?php

error_reporting(E_ALL);
ini_set('display_errors', 1);

include __DIR__ . '/../vendor/autoload.php';


tests/ColorfulTest.php
<?php

declare(strict_types=1);

namespace Jfcherng\Color\Test;

use Jfcherng\Color\Colorful;
use PHPUnit\Framework\TestCase;

/**
 * @covers \Jfcherng\Color\Colorful
 */
class ColorfulTest extends TestCase
{
    /**
     * Provide testcases for testing Colorful::color.
     *
     * @return array the testcases
     */
    public function colorTestcaseProvider(): array
    {
        return [
            [
                [''],
                "\033[0m",
            ],
            [
                ['', 'b_green'],
                "\033[0m",
            ],
            [
                ['', 'b_green', false],
                "\033[42m",
            ],
            [
                ['foo'],
                'foo' . "\033[0m",
            ],
            [
                ['foo' . "\033[0m"],
                'foo' . "\033[0m",
            ],
            [
                ['foo', ''],
                'foo' . "\033[0m",
            ],
            [
                ['foo', 'b_green, b_green'],
                "\033[42m" . 'foo' . "\033[0m",
            ],
            [
                ['foo', ['foo', 'b_green', 'bar'], false],
                "\033[42m" . 'foo',
            ],
            [
                ['foo', 'f_red, b_green'],
                "\033[31;42m" . 'foo' . "\033[0m",
            ],
            [
                ['foo', ['f_red', 'b_green', 'blk']],
                "\033[31;42;5m" . 'foo' . "\033[0m",
            ],
            [
                ['foo', ['f_red', 'b_green', 'blk'], false],
                "\033[31;42;5m" . 'foo',
            ],
        ];
    }

    /**
     * Provide testcases for testing Colorful::noColor.
     *
     * @return array the testcases
     */
    public function noColorTestcaseProvider(): array
    {
        return [
            [
                "\033[31;42;5m" . 'foo' . "\033[0m" . 'bar',
                'foobar',
            ],
        ];
    }

    /**
     * Test Colorful::color.
     *
     * @dataProvider colorTestcaseProvider
     *
     * @param array  $inputs   The inputs
     * @param string $expected The expected
     */
    public function testColor(array $inputs, string $expected): void
    {
        $this->assertSame($expected, Colorful::color(...$inputs));
    }

    /**
     * Test Colorful::noColor.
     *
     * @dataProvider noColorTestcaseProvider
     *
     * @param string $inputs   The input
     * @param string $expected The expected
     */
    public function testNoColor(string $input, string $expected): void
    {
        $this->assertSame($expected, Colorful::noColor($input));
    }
}


src/helpers.php
<?php

declare(strict_types=1);

if (!\function_exists('str_color')) {
    /**
     * Make a string colorful.
     *
     * A global alias to \Jfcherng\Color\Colorful::color.
     *
     * @param string       $str       the string
     * @param array|string $colors    the colors
     * @param bool         $autoReset automatically reset at the end of the string?
     *
     * @return string the colored string
     */
    function str_color(string $str, $colors = [], bool $autoReset = true): string
    {
        return \Jfcherng\Color\Colorful::color($str, $colors, $autoReset);
    }
}

if (!\function_exists('str_nocolor')) {
    /**
     * Remove all colors from a string.
     *
     * A global alias to \Jfcherng\Color\Colorful::noColor
     *
     * @param string $str the string
     *
     * @return string the string without colors
     */
    function str_nocolor(string $str): string
    {
        return \Jfcherng\Color\Colorful::noColor($str);
    }
}


src/Colorful.php
<?php

declare(strict_types=1);

namespace Jfcherng\Color;

/**
 * Make your PHP command-line application colorful.
 *
 * @see https://en.wikipedia.org/wiki/ANSI_escape_code
 *
 * @author Jack Cherng <jfcherng@gmail.com>
 */
class Colorful
{
    const COLOR_BEGIN = "\033[";
    const COLOR_END = 'm';

    const COLOR_BEGIN_REGEX = "\033\\[";
    const COLOR_END_REGEX = 'm';

    /**
     * @var array the color map
     */
    protected static $colorMap = [
        // background
        'b_black' => '40',
        'b_blue' => '44',
        'b_cyan' => '46',
        'b_green' => '42',
        'b_light_gray' => '47',
        'b_magenta' => '45',
        'b_red' => '41',
        'b_yellow' => '43',

        // foreground
        'f_black' => '30',
        'f_blue' => '34',
        'f_brown' => '33',
        'f_cyan' => '36',
        'f_green' => '32',
        'f_light_gray' => '37',
        'f_normal' => '39',
        'f_purple' => '35',
        'f_red' => '31',

        // compound
        'f_dark_gray' => '1;30',
        'f_light_blue' => '1;34',
        'f_light_cyan' => '1;36',
        'f_light_green' => '1;32',
        'f_light_purple' => '1;35',
        'f_light_red' => '1;31',
        'f_white' => '1;37',
        'f_yellow' => '1;33',

        // special
        'blink' => '5',
        'bold' => '1',
        'dim' => '2',
        'hidden' => '8',
        'reset' => '0',
        'reverse' => '7',
        'underline' => '4',

        // alias
        'b' => 'bold',
        'blk' => 'blink',
        'h' => 'hidden',
        'rev' => 'reverse',
        'rst' => 'reset',
        'u' => 'underline',

        // regex for color codes
        'regex_any' => '(?:[0-9]++;?)++',
    ];

    /**
     * Get the color map.
     *
     * @return array the color map
     */
    public static function getColorMap(): array
    {
        return static::$colorMap;
    }

    /**
     * Make a string colorful.
     *
     * @param string       $str       the string
     * @param array|string $colors    the colors
     * @param bool         $autoReset automatically reset at the end of the string?
     *
     * @return string the colored string
     */
    public static function color(string $str, $colors = [], bool $autoReset = true): string
    {
        // always convert $colors into an array
        if (\is_string($colors)) {
            $colors = \explode(',', $colors);
        }

        $colored = static::getColorCode($colors) . $str;

        if ($autoReset) {
            $colored .= static::getColorCode(['reset']);
        }

        return static::simplifyColoredString($colored);
    }

    /**
     * Remove all colors from a string.
     *
     * @param string $str the string
     *
     * @return string the string without colors
     */
    public static function noColor(string $str): string
    {
        return \preg_replace(
            '~' . static::getColorCode(['regex_any'], true) . '~uS',
            '',
            $str
        );
    }

    /**
     * Get the color code from given colors.
     *
     * @param array $colors      the colors
     * @param bool  $returnRegex return as an regex segment
     *
     * @return string the color code
     */
    protected static function getColorCode(array $colors, bool $returnRegex = false): string
    {
        $colors = static::sanitizeColors($colors);

        if (empty($colors)) {
            return '';
        }

        $colorCodes = \array_map(
            function (string $color): string {
                while (isset(static::$colorMap[$color])) {
                    $color = static::$colorMap[$color];
                }

                return $color;
            },
            $colors
        );

        $closures = $returnRegex
            ? [static::COLOR_BEGIN_REGEX, static::COLOR_END_REGEX]
            : [static::COLOR_BEGIN, static::COLOR_END];

        return $closures[0] . \implode(';', $colorCodes) . $closures[1];
    }

    /**
     * Sanitize colors.
     *
     * @param array $colors the colors
     *
     * @return array the sanitized colors
     */
    protected static function sanitizeColors(array $colors): array
    {
        return \array_unique(\array_filter(
            \array_map('trim', $colors),
            function (string $color): bool {
                return isset(static::$colorMap[$color]);
            }
        ));
    }

    /**
     * Simplify the colored string.
     *
     * @param string $str the colored string
     *
     * @return string the simplified colored string
     */
    protected static function simplifyColoredString(string $str): string
    {
        // replace multiple consecutive resets with a single reset
        $str = \preg_replace(
            '~(' . static::getColorCode(['reset'], true) . '){2,}~uS',
            '$1',
            $str
        );

        // remove colors for an emtpy string
        $str = \preg_replace(
            (
                '~' .
                    '(' . static::getColorCode(['regex_any'], true) . ')' .
                    '(' . static::getColorCode(['reset'], true) . ')' .
                '~uS'
            ),
            '$2',
            $str
        );

        return $str;
    }
}
